File: jaenny/Test/220508 우테캠/3.js
let matrix = [];
let n = 0;
let m = 0;

function solution(board) {
  var answer = 0;
  matrix = board;
  n = matrix.length;
  m = matrix[0].length;

  // 세로
  for (let j = 0; j < m; j++) {
    let visit = createVisit(n, m);
    let res = bfs(0, j, 'down', visit);
    // console.log(j, res);
    if (answer < res) answer = res;
  }

  // 가로
  for (let i = 0; i < n; i++) {
    let visit = createVisit(n, m);
    let res = bfs(i, 0, 'right', visit);
    // console.log(i, res);
    if (answer < res) answer = res;
  }

  return answer;
}

function bfs(i, j, direction, visit) {
  const dy = [1, -1, 0, 0];
  const dx = [0, 0, 1, -1];

  let len = direction == 'down' ? visit.length : visit[0].length;

  let cnt = 0;
  let queue = [];

  if (direction == 'down') {
    for (let i = 0; i < len; i++) {
      if (visit[i][j] != 1) {
        visit[i][j] = 1;
        queue.push([i, j]);
        target = matrix[i][j];
        cnt += 1;

        while (queue) {
          if (queue.length == 0) break;
          // console.log(queue);
          let node = queue.shift();
          // console.log(node);
          let y = node[0];
          let x = node[1];

          for (let d = 0; d < 4; d++) {
            let ny = y + dy[d];
            let nx = x + dx[d];

            if (
              ny >= 0 &&
              ny < n &&
              nx >= 0 &&
              nx < m &&
              visit[ny][nx] != 1 &&
              matrix[ny][nx] == target
            ) {
              visit[ny][nx] = 1;
              cnt += 1;
              queue.push([ny, nx]);
            }
          }
        }
      }
    }
  } else {
    for (let j = 0; j < len; j++) {
      if (visit[i][j] != 1) {
        visit[i][j] = 1;
        queue.push([i, j]);
        target = matrix[i][j];
        cnt += 1;

        while (queue) {
          if (queue.length == 0) break;
          let node = queue.shift();
          let y = node[0];
          let x = node[1];

          for (let d = 0; d < 4; d++) {
            let ny = y + dy[d];
            let nx = x + dx[d];

            if (
              ny >= 0 &&
              ny < n &&
              nx >= 0 &&
              nx < m &&
              visit[ny][nx] != 1 &&
              matrix[ny][nx] == target
            ) {
              visit[ny][nx] = 1;
              cnt += 1;
              queue.push([ny, nx]);
            }
          }
        }
      }
    }
  }

  return cnt;
}

function createVisit(rows, cols) {
  let arr = new Array(rows);
  for (let i = 0; i < rows; i++) {
    arr[i] = new Array(cols);
  }
  return arr;
}

console.log(
  solution(['ABBBBC', 'AABAAC', 'BCDDAC', 'DCCDDE', 'DCCEDE', 'DDEEEB'])
);
console.log(solution(['DDCCC', 'DBBBC', 'DBABC', 'DBBBC', 'DDCCC']));

File: jaenny/Test/220508 우테캠/3.py
from collections import deque
matrix = []
n = 0
m = 0
def solution(board) :
  answer = 0
  global matrix, n, m
  matrix = board
  n = len(matrix)
  m = len(matrix[0])


  # 세로
  
  for j in range(m) :
    visit = [[0]*m for _ in range(n)]
    r = bfs(0,j,'down',visit)
    answer = max(answer, r)
    print(j, r)
    print('------------')
  
  for i in range(n) :
    visit = [[0]*m for _ in range(n)]
    r = bfs(i,0,'right',visit)
    answer = max(answer, r)
    print(i, r)
    print('------------')
  
  return answer

dy = [1,-1,0,0]
dx = [0,0,1,-1]

def bfs(i,j,direction,visit) :
  cnt = 0
  queue = deque()
  if direction == 'down' :
    for i in range(len(visit)) :
      if visit[i][j] == 0 :
        visit[i][j] = 1
        queue.append([i,j])
        target = matrix[i][j]
        cnt += 1
        while queue :
          print(queue)
          y,x = queue.popleft()

          for d in range(4) :
            ny = y + dy[d]
            nx = x + dx[d]
            if 0<= ny < n and 0<= nx < m and visit[ny][nx] == 0 and matrix[ny][nx] == target :
              visit[ny][nx] = 1
              cnt += 1
              queue.append([ny,nx])
  else :
    for j in range(len(visit[0])) :
      if visit[i][j] == 0 :
        visit[i][j] = 1
        queue.append([i,j])
        target = matrix[i][j]
        cnt += 1
        while queue :
          y,x = queue.popleft()

          for d in range(4) :
            ny = y + dy[d]
            nx = x + dx[d]
            if 0<= ny < n and 0<= nx < m and visit[ny][nx] == 0 and matrix[ny][nx] == target :
              visit[ny][nx] = 1
              cnt += 1
              queue.append([ny,nx])
  
  return cnt



# print(solution(["DDCCC","DBBBC","DBABC","DBBBC","DDCCC"]))
print(solution(["ABBBBC","AABAAC","BCDDAC","DCCDDE","DCCEDE","DDEEEB"]))